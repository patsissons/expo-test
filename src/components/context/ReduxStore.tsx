import * as React from 'react';
import { Reducer, Middleware, StoreEnhancer, createStore, applyMiddleware, Dispatch, bindActionCreators } from 'redux';
import { Provider, connect } from 'react-redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { Store, ContextState } from '../state';
import { Actions, actions } from '../actions';
import { getTypeNames, isDebugEnvironment } from '../../utils';

export namespace ReduxProvider {
  export interface StateProps<T> {
    state: T;
  }

  export interface ActionProps {
    actions: Actions;
  }

  export interface ContextProps {
    context: ContextState;
  }

  export interface Props {
    reducers: Reducer<any>;
    epicMiddleware: Middleware;
    reactNavigationMiddleware: Middleware;
  }

  export interface ComponentProps<T> extends StateProps<T>, ActionProps, ContextProps {
  }
}

export class ReduxProvider extends React.Component<ReduxProvider.Props> {
  render() {
    const middleware = applyMiddleware(
      this.props.epicMiddleware,
      this.props.reactNavigationMiddleware,
    );

    let storeEnhancer: StoreEnhancer<any> = middleware;

    // istanbul ignore next impossible code path when testing
    if (isDebugEnvironment()) {
      storeEnhancer = composeWithDevTools(middleware);
    }

    const store = createStore(
      this.props.reducers,
      storeEnhancer,
    );

    return (
      <Provider store={ store }>
        { this.props.children }
      </Provider>
    );
  }
}

function mapDispatchToActions(dispatch: Dispatch<Store>): ReduxProvider.ActionProps {
  return {
    actions: getTypeNames(actions)
      .reduce(
        (map, actionName) => {
          map[actionName] = bindActionCreators(actions[actionName], dispatch);

          return map;
        },
        {} as any,
      ),
  };
}

function mapStateToProps<TState>(
  store: Store,
  stateSelector: (store: Store) => TState,
) {
  return {
    state: stateSelector(store),
    context: store.context,
  };
}

export function withState<T extends ReduxProvider.ComponentProps<TState>, TState>(
  Component: React.ComponentType<T>,
  stateSelector: (store: Store) => TState,
) {
  const connector = connect(
    (store: Store) => mapStateToProps(store, stateSelector),
    mapDispatchToActions,
  );

  return connector(Component);
}

export function withStore<T extends ReduxProvider.ComponentProps<Store>>(
  Component: React.ComponentType<T>,
) {
  return withState(Component, x => x);
}
