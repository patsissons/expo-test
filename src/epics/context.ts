import { Observable } from 'rxjs';
import { ActionsObservable } from 'redux-observable';
import { DangerZone } from 'expo';
import { RequestAction, actionTypes, getRequestAction, LocalePayload } from '../framework';

export function loadLocale(actions: ActionsObservable<RequestAction>): Observable<RequestAction<LocalePayload>> {
  return actions
    .ofType(actionTypes.context.LoadLocale)
    .flatMap(() => Observable.from(DangerZone.Localization.getCurrentLocaleAsync()))
    .map(locale => getRequestAction(actionTypes.context.LoadLocalization, { locale }));
}
