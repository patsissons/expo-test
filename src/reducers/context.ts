import { ContextState, RequestAction, LocalePayload, isLocalePayload, actionTypes } from '../components';
import { getLocalization } from '../locale';
import { env } from '../utils';

const initialState: Partial<ContextState> = {
};

export function reducer(state = initialState, action: RequestAction<LocalePayload>): Partial<ContextState> {
  switch (action.type) {
    case actionTypes.context.LoadEnv:
      return {
        ...state,
        env,
      };
    case actionTypes.context.LoadLocalization:
      if (isLocalePayload(action.payload)) {
        return {
          ...state,
          locale: action.payload.locale,
          localization: getLocalization(action.payload.locale),
        };
      }
      break;
    default:
      // fall through and return input state
      break;
  }

  return state;
}
