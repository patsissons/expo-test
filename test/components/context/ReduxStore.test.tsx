import React from 'react';
import { shallow } from 'enzyme';
import * as redux from 'redux';
import { ReduxProvider } from '../../../src/framework/context/ReduxStore';
import {  } from 'redux-devtools-extension';

const applyMiddleware = redux.applyMiddleware as jest.Mock;
const createStore = redux.createStore as jest.Mock;

jest.mock('redux', () => {
  const store = {
    subscribe: jest.fn(),
    dispatch: jest.fn(),
    getState: jest.fn(),
  };

  return {
    applyMiddleware: jest.fn(() => ({})),
    createStore: jest.fn(() => store),
  };
});

describe('components', () => {
  describe('context', () => {
    describe('ReduxProvider', () => {
      it('renders without crashing', () => {
        // Arrange
        const props = {
          reducers: {},
          epicMiddleware: {},
          reactNavigationMiddleware: {},
        };
        const component = (<ReduxProvider { ...props as any }><div /></ReduxProvider>);

        // Act
        const result = shallow(component);

        // Assert
        expect(result).toMatchSnapshot();
      });

      it('injects a redux store', () => {
        // Arrange
        const middleware = {};
        const store = createStore();
        const props = {
          reducers: {},
          epicMiddleware: {},
          reactNavigationMiddleware: {},
        };
        applyMiddleware.mockImplementation(() => middleware);
        createStore.mockImplementation(() => store);
        const component = (<ReduxProvider { ...props as any }><div /></ReduxProvider>);

        // Act
        const result = shallow(component);

        // Assert
        expect(applyMiddleware).toHaveBeenCalledWith(props.epicMiddleware, props.reactNavigationMiddleware);
        expect(createStore).toHaveBeenCalledWith(props.reducers, middleware);
        expect(result.prop('store')).toBe(store);
      });
    });
  });
});
