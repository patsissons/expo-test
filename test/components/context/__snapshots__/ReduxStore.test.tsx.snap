// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components context ReduxProvider renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReduxProvider
    epicMiddleware={Object {}}
    reactNavigationMiddleware={Object {}}
    reducers={Object {}}
  >
    <div />
  </ReduxProvider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <div />,
      "store": Object {
        "dispatch": [MockFunction],
        "getState": [MockFunction],
        "subscribe": [MockFunction],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <div />,
        "store": Object {
          "dispatch": [MockFunction],
          "getState": [MockFunction],
          "subscribe": [MockFunction],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
